### PR TITLE
Add adr-002 evm hooks proposal

### PR DESCRIPTION
**Pull request by yihuang**
## Description

ref impl: https://github.com/tharsis/ethermint/pull/417

______

For contributor use:

- [x] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [x] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added  relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer

______

For admin use:

- [ ] Added appropriate labels to PR (ex. `WIP`, `R4R`, `docs`, etc)
- [ ] Reviewers assigned
- [ ] Squashed all commits, uses message "Merge pull request #XYZ: [title]" ([coding standards](https://github.com/tendermint/coding/blob/master/README.md#merging-a-pr))

### CODE REVIEW

File: docs/architecture/adr-002-evm-hooks.md [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+This ADR propose to add `EvmHooks` interface:
+
+```golang
+type EvmHooks interface {
+	  PostTxProcessing(ctx sdk.Context, txHash ethcmn.Hash, logs []*ethtypes.Log) error
+}
+```
+
+- `PostTxProcessing` is called after EVM transaction finished, executed with the same cache context as the EVM
```
[the remark this answers is not in the document]
Comment (yihuang) [reply docs/architecture/adr-002-evm-hooks.md:31]: done, added the bank send example.

File: docs/architecture/adr-002-evm-hooks.md [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+There are some requirements to enable EVM contract to communicate with cosmos native modules, which could have multiple
+use cases. We need some kind of hooks to support these extension requirements in a generic way.
+
+## Decision
+
+This ADR propose to add `EvmHooks` interface:
+
+```golang
+type EvmHooks interface {
+	  PostTxProcessing(ctx sdk.Context, txHash ethcmn.Hash, logs []*ethtypes.Log) error
```
[the remark this answers is not in the document]
Comment (yihuang) [reply docs/architecture/adr-002-evm-hooks.md:27]: yeah, it's possible, it's just I don't see an immediately use case, maybe someone could added in another proposal when actual use case appeared.

[the remark this answers is not in the document]
Comment (fedekunze) [reply docs/architecture/adr-002-evm-hooks.md:27]: the implementation in #417 doesn't receive the context. Why the difference?

File: docs/architecture/adr-002-evm-hooks.md [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+}
+```
+
+- `PostTxProcessing` is called after EVM transaction finished, executed with the same cache context as the EVM
+  transaction execution, if `PostTxProcessing` returns an error, the whole EVM transaction is reverted.
```
Comment (fedekunze) [inline docs/architecture/adr-002-evm-hooks.md:32]: explain why are you reverting the entire tx vs just doing a no-op for the hook post tx processing

File: docs/architecture/adr-002-evm-hooks.md [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+- 2021-08-11: first draft
+
+## Status
+
+PROPOSED
+
+## Abstract
+
+The current ADR proposes a hook interface to EVM module, with the goal of extend the functionality externally,
```
Comment (fedekunze) [inline docs/architecture/adr-002-evm-hooks.md:13]: > the functionality

explain which functionality

File: docs/architecture/adr-002-evm-hooks.md [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+specifically to support communication from EVM contract to native modules through logs.
+
+## Context
+
+There are some requirements to enable EVM contract to communicate with cosmos native modules, which could have multiple
+use cases. We need some kind of hooks to support these extension requirements in a generic way.
```
Comment (fedekunze) [inline docs/architecture/adr-002-evm-hooks.md:19]: explain use cases

File: docs/architecture/adr-002-evm-hooks.md [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+## Status
+
+PROPOSED
+
+## Abstract
+
+The current ADR proposes a hook interface to EVM module, with the goal of extend the functionality externally,
+specifically to support communication from EVM contract to native modules through logs.
+
+## Context
```
Comment (fedekunze) [inline docs/architecture/adr-002-evm-hooks.md:16]: This requires a more descriptive paragraph of the problem and the limitations of the current implementation.

> This section describes the forces at play, including technological, political, social, and project local. These forces are probably in tension, and should be called out as such. The language in this section is value-neutral. It is simply describing facts. It should clearly explain the problem and motivation that the proposal aims to resolve.

File: docs/architecture/adr-002-evm-hooks.md [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+  }
+  ```
+
+There are no hooks implemented in the default application, but other applications could implement custom hooks and
+register them to the `EvmKeeper`, for example:
+
+```golang
+evmKeeper.SetHooks(NewHook());
+```
+
```
Comment (fedekunze) [inline docs/architecture/adr-002-evm-hooks.md:52]: There are a bunch of sections missing here:

- Backwards Compatibility
- Positive
- Negative 
- Neutral
- Further Discussions
- Test Cases
- References

**Review by fedekunze — CHANGES_REQUESTED** [review]
The ADR is good, but it doesn't follow the ADR template and a lot of context and details are missing if you compare it with the implementation. For example, the following decision is not documented on the ADR:

```
if !res.Failed() {
	err = k.PostTxProcessing(txHash, logs)
	if err != nil {
		k.RevertToSnapshot(revision)
		res.VmError = "failed to process native logs"
		res.Ret = []byte(err.Error())
	}
}
```

Some questions that need to be documented:

- why do we revert the state to the snapshot prior to the tx execution? (i.e why did you choose this approach over just performing a no-op on `PostTxProcessing`)
- give a reference implementation for a `Hook` implementation
- explain the `MultiHook` design for adding multiple hooks
- explain how to add the Hook to the `app.go`

File: docs/architecture/adr-002-evm-hooks.md (hunk 3 above)
Comment (yihuang) [reply docs/architecture/adr-002-evm-hooks.md]: added.
it makes the implementor to decide if certain api want to revert on error or not.

File: docs/architecture/adr-002-evm-hooks.md (hunk 4 above)
Comment (yihuang) [reply docs/architecture/adr-002-evm-hooks.md]: changed to tx processing logic.

File: docs/architecture/adr-002-evm-hooks.md (hunk 5 above)
Comment (yihuang) [reply docs/architecture/adr-002-evm-hooks.md]: Added more detail to the "Call Native Module" use case.

File: docs/architecture/adr-002-evm-hooks.md (hunk 6 above)
Comment (yihuang) [reply docs/architecture/adr-002-evm-hooks.md:16]: Added some explanation for the motives to add hook interface.

File: docs/architecture/adr-002-evm-hooks.md (hunk 7 above)
Comment (yihuang) [reply docs/architecture/adr-002-evm-hooks.md:138]: added

**Comment by yihuang** [comment]
> The ADR is good, but it doesn't follow the ADR template and a lot of context and details are missing if you compare it with the implementation. For example, the following decision is not documented on the ADR:
> 
> ```
> if !res.Failed() {
> 	err = k.PostTxProcessing(txHash, logs)
> 	if err != nil {
> 		k.RevertToSnapshot(revision)
> 		res.VmError = "failed to process native logs"
> 		res.Ret = []byte(err.Error())
> 	}
> }
> ```

added a little bit more explanation to this:
- hooks are called only when tx executed successfully
- errors returned by hooks get translated to vm error, and cause the whole tx to revert.

> 
> Some questions that need to be documented:
> 
> * why do we revert the state to the snapshot prior to the tx execution? (i.e why did you choose this approach over just performing a no-op on `PostTxProcessing`)

added, with this approach, the implementor is still free to swallow any error.

> * give a reference implementation for a `Hook` implementation

add the `BankSendHook` to adr.

> * explain the `MultiHook` design for adding multiple hooks

I think this is not essential for the proposal?

> * explain how to add the Hook to the `app.go`

added.

File: docs/architecture/adr-002-evm-hooks.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (yihuang) [reply docs/architecture/adr-002-evm-hooks.md:34]: > the implementation in #417 doesn't receive the context. Why the difference?

In newer version, it's the same as the adr.

**Review by fedekunze — COMMENTED** [review]
Looks better now, although there are some details and code snippet are missing about the integration with the `state_transition.go`. 

I will also recommend you use [Grammarly](https://www.grammarly.com/) to fix some of the wording.

File: docs/architecture/adr-002-evm-hooks.md [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+func NewBankSendHook(bankKeeper bankkeeper.Keeper) *BankSendHook {
+	return &BankSendHook{
+		bankKeeper: bankKeeper,
+	}
+}
+
+func (h BankSendHook) PostTxProcessing(ctx sdk.Context, txHash ethcmn.Hash, logs []*ethtypes.Log) error {
+	for _, log := range logs {
+		if len(log.Topics) > 0 && log.Topics[0] == BankSendEvent.ID {
```
[the remark this answers is not in the document]
Comment (yihuang) [reply docs/architecture/adr-002-evm-hooks.md:117]: You mean whitelisting contracts that are allowed to do this call? that may be a business decision to make, as long as the call is secure, I don't see why we shouldn't open it to all contracts.

[the remark this answers is not in the document]
Comment (fedekunze) [reply docs/architecture/adr-002-evm-hooks.md:117]: > that may be a business decision to make, as long as the call is secure, I don't see why we shouldn't open it to all contracts.

for example: what if someone creates an event with the same name that doesn't deduct the funds from the user? 

In general, I think it's ok to leave the assumption that the application will be in charge of the security of the execution of state changes with regards to using logs in conjunction with SDK modules, although we should definitely explain that that on the `Neutral` consequences with a warning.

[the remark this answers is not in the document]
Comment (fedekunze) [reply docs/architecture/adr-002-evm-hooks.md:117]: also add the comment on the ADR above the `if len(log.Topics) == 0 || log.Topics[0] != BankSendEvent.ID` condition

[the remark this answers is not in the document]
Comment (yihuang) [reply docs/architecture/adr-002-evm-hooks.md:117]: Yeah, I agree that the potential name conflict should be put into `Neutral` or even `Negative` consequence.
I'll add an item into `Negative` consequence, and add comments in the code.

File: docs/architecture/adr-002-evm-hooks.md [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+## Further Discussions
+
+<!-- While an ADR is in the DRAFT or PROPOSED stage, this section should contain a summary of issues to be solved in future iterations (usually referencing comments from a pull-request discussion).
+Later, this section can optionally list ideas or improvements the author or reviewers found during the analysis of this ADR. -->
+
+## Test Cases [optional]
+
+<!-- Test cases for an implementation are mandatory for ADRs that are affecting consensus changes. Other ADRs can choose to include links to test cases if applicable. -->
+
+## References
```
[the remark this answers is not in the document]
Comment (yihuang) [reply docs/architecture/adr-002-evm-hooks.md:180]: done

File: docs/architecture/adr-002-evm-hooks.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (yihuang) [reply docs/architecture/adr-002-evm-hooks.md]: done

File: docs/architecture/adr-002-evm-hooks.md [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+	}
+}
+
+func (h BankSendHook) PostTxProcessing(ctx sdk.Context, txHash ethcmn.Hash, logs []*ethtypes.Log) error {
+	for _, log := range logs {
+		if len(log.Topics) > 0 && log.Topics[0] == BankSendEvent.ID {
+			unpacked, err := BankSendEvent.Inputs.Unpack(log.Data)
+			if err != nil {
+				// ignore unrecognizable log
+				continue
```
[the remark this answers is not in the document]
Comment (yihuang) [reply docs/architecture/adr-002-evm-hooks.md]: done

**Comment by yihuang** [comment]
> Looks better now, although there are some details and code snippet are missing about the integration with the `state_transition.go`.

done, added some `ApplyTransaction` code snippet.

> 
> I will also recommend you use [Grammarly](https://www.grammarly.com/) to fix some of the wording.

done

**Review by fedekunze — APPROVED** [review]
ACK 👍

**Pull request merged**